meps/mep-cross-references.md

```diff
@@ -0,0 +1,339 @@
+- `[My cool fig](my-fig-ref)` - labeled reference
+- `` {numref}`my-fig-ref` `` - numbered reference (e.g. "Fig. 1")
+- `` {numref}`Custom Figure %s text <my-fig-ref>` `` - custom numbering for the figure, using `%s` or `{number}`
+
+**Equations**
+
+- `` {eq}`my-eqn` `` - uses a custom role, only for equations, no ability to override title
+- `` {math:numref}`my-eqn` `` - a verbose form of the `eq` role (see note about "Domains" below).
+- `[](my-math-ref)` - using existing MyST link syntax (note this does not have a `#`)
+
+**Documents**
+
+- `` {doc}`my-eqn` `` - a doc role.
+- `[](../file-types/myst-notebooks.md)` - a relative path with POSIX path separators `/`, if no title provided MyST will fill in with the referenced title (or file name).
+- `[A different page](../file-types/myst-notebooks.md)` - using a custom title.
+
+**Intersphinx**
+
+Multiple other sphinx documentation sites can be referenced in MyST syntax ([Sphinx documentation](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#module-sphinx.ext.intersphinx)). For example, the `python` documentation can be referenced from a configuration (e.g. the [intersphinx_mapping](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#configuration) in `conf.py`), which points to the appropriate intersphinx inventory (e.g. `https://docs.python.org/3`) containing a `*.inv` file.
+
+- `` {external+python:py:class}`zipfile.ZipFile` `` - a reference to the Python class `ZipFile`
+- `` {py:class}`zipfile.ZipFile` `` - a short-hand reference that will search locally first, then any referenced inventories
+
+:::{note}
+Although we consider intersphinx in this MEP, we do not suggest a specific syntax, and will leave that discussion to a future MEP.
+:::
+
+**Styling**
+
+All link syntax supports styling inside of the reference, (e.g. `[A **bolded _reference_** to a page](./myst.md)`) the reference role syntax currently does **not** support styling of the inner content.
+
+## Proposal
+
+We propose a cross-reference syntax that uses CommonMark links in all three forms. The goal is to support all use cases of cross-referencing with the most common use cases of referencing a document, file, section or element being simple, terse and familiar.
+
+**Overview:**
+
+| Existing Syntax                          | New Syntax                   |
+| :--------------------------------------- | :--------------------------- |
+| `[](my-id)`[^legacy_hash]                | `[](#my-id)`                 |
+| `` {ref}`my-id` ``                       | `[](#my-id)`                 |
+| `` {eq}`my-equation` ``                  | `[](#my-equation)`           |
+| `` {ref}`Custom Text <my-id>` ``         | `[Custom Text](#my-id)`      |
+| `` {ref}`See "{name}" <my-id>` ``        | `[See "%t"](#my-id)`         |
+| `` {numref}`Custom Number %s <my-id>` `` | `[Custom Number %s](#my-id)` |
+| `` {doc}`my-doc` ``                      | `[](my-doc.md)`              |
```

Comment:
In the new syntax, it looks like you always need to provide the extensions when referencing other pages? Is there any implicit resolution of `[](my-doc)`? I could see this getting confusing if source files have different extensions (`md`, `ipynb`, `tex`...), particularly in external projects. (Maybe that's getting into complexity with intersphinx, but still seems relevant within a single project.)

Reply:
Resolved in a comment below!